crates/bevy_pbr/src/light_probe/generate.rs

```diff
@@ -0,0 +1,1084 @@
+    render_graph::{Node, NodeRunError, RenderGraphContext, RenderLabel},
+    render_resource::{
+        binding_types::*, AddressMode, BindGroup, BindGroupEntries, BindGroupLayout,
+        BindGroupLayoutEntries, CachedComputePipelineId, ComputePassDescriptor,
+        ComputePipelineDescriptor, Extent3d, FilterMode, PipelineCache, Sampler,
+        SamplerBindingType, SamplerDescriptor, ShaderDefVal, ShaderStages, ShaderType,
+        StorageTextureAccess, Texture, TextureAspect, TextureDescriptor, TextureDimension,
+        TextureFormat, TextureSampleType, TextureUsages, TextureView, TextureViewDescriptor,
+        TextureViewDimension, UniformBuffer,
+    },
+    renderer::{RenderContext, RenderDevice, RenderQueue},
+    settings::WgpuFeatures,
+    sync_world::RenderEntity,
+    texture::{CachedTexture, GpuImage, TextureCache},
+    Extract,
+};
+
+use bevy_light::{EnvironmentMapLight, GeneratedEnvironmentMapLight};
+use core::cmp::min;
+
+/// Handle for Spatio-Temporal Blue Noise texture
+pub const STBN: Handle<Image> = uuid_handle!("3110b545-78e0-48fc-b86e-8bc0ea50fc67");
+
+/// Labels for the environment map generation nodes
+#[derive(PartialEq, Eq, Debug, Copy, Clone, Hash, RenderLabel)]
+pub enum GeneratorNode {
+    Mipmap,
+    Radiance,
+    Irradiance,
+}
+
+/// Stores the bind group layouts for the environment map generation pipelines
+#[derive(Resource)]
+pub struct GeneratorBindGroupLayouts {
+    pub spd: BindGroupLayout,
+    pub radiance: BindGroupLayout,
+    pub irradiance: BindGroupLayout,
+    pub copy: BindGroupLayout,
+}
+
+/// Samplers for the environment map generation pipelines
+#[derive(Resource)]
+pub struct GeneratorSamplers {
+    pub linear: Sampler,
+}
+
+/// Pipelines for the environment map generation pipelines
+#[derive(Resource)]
+pub struct GeneratorPipelines {
+    pub spd_first: CachedComputePipelineId,
+    pub spd_second: CachedComputePipelineId,
+    pub radiance: CachedComputePipelineId,
+    pub irradiance: CachedComputePipelineId,
+    pub copy: CachedComputePipelineId,
+}
+
+/// Initializes all render-world resources used by the environment-map generator once on
+/// [`bevy_render::RenderStartup`].
+pub fn init_generator_resources(
+    mut commands: Commands,
+    render_device: Res<RenderDevice>,
+    pipeline_cache: Res<PipelineCache>,
+    asset_server: Res<AssetServer>,
+) {
+    // Bind group layouts
+    let spd = render_device.create_bind_group_layout(
+        "spd_bind_group_layout",
+        &BindGroupLayoutEntries::with_indices(
+            ShaderStages::COMPUTE,
+            (
+                (
+                    0,
+                    texture_2d_array(TextureSampleType::Float { filterable: true }),
+                ), // Source texture
+                (
+                    1,
+                    texture_storage_2d_array(
+                        TextureFormat::Rgba16Float,
+                        StorageTextureAccess::WriteOnly,
+                    ),
```

Comment:
this could probably be a variable pulled out, seems to be the same for all of the mips. Can you just clone it a bunch of times instead?

Reply:
resolved by https://github.com/mate-h/bevy/pull/8